### PR TITLE
Show active dispute as inline notice in order edit screen

### LINKED ISSUE #6354: Epic: WCPay Disputes – Highlight active dispute in Woo order details screen
### Description
The goal of this epic is to improve visibility of active disputes in the Woo orders screen, by adding a new notice (and improving clarity of the order note).

Screenshot 2023-05-23 at 1 08 42 PM

order-details-screen

⚠️ **Note** – the wording of the notice/note in the screenshot is out of date. Updated wording below.

* Colour and wording:
	* \> 7 days – 🟡 Yellow, `This order has a chargeback dispute of {AMOUNT} labeled as {REASON}. Please respond to this dispute before {DEADLINE}`
	* < 7 days – 🟡 Yellow, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
	* < 72hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
	* < 24hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. (Last day today)`

### Steps
- [x] Clearer & more informative order note when order is disputed - see prototype / WIP PRs:
  - https://github.com/woocommerce/woocommerce/pull/38443
  - #6535
  - #6373
- [x] #6379 
- [x] Register Tracks events
	- [x] #6742
- [x] Add testing instructions to this description for impacted flows

The above is a high-level summary of the changes in the design. These can be split off as separate issues. Can add / adapt if details change.

### Additional context (internal links)
- Project thread: pdjTHR-2Hv-p2

## Testing flows

- Install and activate WC 7.9+
- Set up a dispute by purchasing with a [dispute test card](https://stripe.com/docs/testing#disputes).
- View the order edit page and confirm the alert shows correct info and button links to dispute detail. 
- Try various combinations of dispute amount, date, and store settings (e.g. multicurrency) to ensure the notice appears when necessary and renders correctly.
- Ensure the notice is not displayed if the dispute has been lost or otherwise resolved, or the merchant has already submitted evidence (responded).

### PR DESCRIPTION
**Pull request by haszari**
Fixes: #6354

See also: https://github.com/woocommerce/woocommerce/pull/38443

#### Changes proposed in this Pull Request

This PR uses a new hook `woocommerce_admin_order_data_after_payment_info` to add a dispute notice to order edit page, inside the order details box. The tracks event `wcpay_order_dispute_notice_action_click` will be called when clicking the notice action button.

* ~It also uses the WCPay `InlineNotice` component which is a stand-in / prototype for a component proposed for WC core.~
	* **Update:** the component we're looking to utilise for this design has already been implemented as `BannerNotice` in #5814.
* Colour and wording **Note these differ from the design below:**
	* \> 7 days – 🟡 Yellow, `This order has a chargeback dispute of {AMOUNT} labeled as {REASON}. Please respond to this dispute before {DEADLINE}`
	* < 7 days – 🟡 Yellow, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
	* < 72hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
	* < 24hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. (Last day today)`

image

#### Screenshots of this PR's implementation
\> 7 days
image

< 7 days
image

< 72 hours
image

< 24 hours
image

 
## Next steps
- [x] Preferably, move changes to `InlineNotice` to a new PR to simplify testing and review.
See #6535.
- [x] Confirm `BannerNotice` is the correct component to utilise for this design.
- [x] Render date and do date logic 100% on client. I'm rendering on the server, but the date should be passed down in binary / unformatted, and then the client can do all the rendering and urgency calculations. This is not absolutely essential, but seems best to me. Note that `moment` is throwing an error due to the current sloppy approach (`value provided is not in a recognized RFC2822 or ISO format`).
   - **Note** from @jinksi: we'll be sharing more logic across FE components (e.g. `isDisputeUrgent()` to determine yellow or red colour) in Phase 2, so lean towards sharing in FE where suitable. See paJDYF-9oX-p2
- [x] Only show notice when there is an active dispute. I haven't tested this, will need to ensure notice is not shown when merchant has responded (etc).
- [x] Test with WC < 7.9 (before hook `woocommerce_admin_order_data_after_payment_info` added)

###### Follow ups
These tasks don't block this PR but would be good to follow up (e.g. log issues).

- [ ] When working on this I discovered that we have a few hard-coded `/ 100` (stripe value => currency) in the code base. These should use a convert utility that knows about zero decimal currencies. Maybe also `* 100`, see also specific issue #4345.

#### Testing instructions

- Install and activate WC 7.9+
- Set up a dispute by purchasing with a [dispute test card](https://stripe.com/docs/testing#disputes).
- View the order edit page and confirm the alert shows correct info and button links to dispute detail. 
- Try various combinations of dispute amount, date, and store settings (e.g. multicurrency) to ensure the notice appears when necessary and renders correctly.
- Ensure the notice is not displayed if the dispute has been lost or otherwise resolved, or the merchant has already submitted evidence (responded).

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6373` or branch name `try/contextual-notice-in-order-details-wccore-hook` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=try/contextual-notice-in-order-details-wccore-hook) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 335ea6e7481083740044c30bffdb5f8e6a02f100
- Build time: 2023-07-12 03:27:48 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by haszari** [comment]
Reviewers – tagged for review to get feedback on the core proposal. No need to thoroughly review this PR (it's a hack demo).

**Comment by haszari** [comment]
Removing review request – this is in progress :) 

Next steps:

- Adapt to using [WCPay `InlineNotice`](https://github.com/Automattic/woocommerce-payments/tree/develop/client/components/inline-notice) (wrapper for [WP core/gutenberg `Notice`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/notice/index.tsx) with some styling tweaks)
- Implement logic to show notice on disputed orders
- Date logic to control alert colour/urgency and `3 days left` suffix

Then we should be good to go :)

**Comment by Jinksi** [comment]
**Update:** the component we're looking to utilise for this design has been implemented as `BannerNotice` (not `InlineNotice`) in #5814. See screenshot using `BannerNotice` for the Order Dispute Notice:

image

**Comment by Jinksi** [comment]
Rua mentioned the following in his todo notes in the PR description:

> Render date and do date logic 100% on client. I'm rendering on the server, but the date should be passed down in binary / unformatted, and then the client can do all the rendering and urgency calculations. 

I'd lean more towards keeping as much logic in client as reasonable – we'll be sharing more logic across FE components (e.g. `isDisputeUrgent()` to determine yellow or red colour) in Phase 2. See paJDYF-9oX-p2

File: changelog/try-contextual-notice-in-order-details-wccore-hook
Comment (shendy-a8c) [inline changelog/try-contextual-notice-in-order-details-wccore-hook]: More clear to users if we use WooCommerce instead of WC.

```suggestion
Add dispute notice to the WooCommerce order screen to highlight disputes awaiting a response.
```

**Comment by Jinksi** [comment]
Note: the `wcpay_order_dispute_notice_action_click` tracks event needs to be registered after this is merged.

**Comment by Jinksi** [comment]
Based on our discussion in Slack (p1689046417627739/1688655548.302669-slack-C03LA109BKM) I have now updated the colour and wording to the following (see screenshots in description):

* \> 7 days – 🟡 Yellow, `This order has a chargeback dispute of {AMOUNT} labeled as {REASON}. Please respond to this dispute before {DEADLINE}`
* < 7 days – 🟡 Yellow, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
* < 72hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. ({N} days left)`
* < 24hours – 🔴 Red, `Please resolve the dispute on this order for {AMOUNT} labeled {REASON} by {DEADLINE}. (Last day today)`

cc @shendy-a8c

File: includes/class-wc-payments-order-service.php
Comment (shendy-a8c) [inline includes/class-wc-payments-order-service.php]: `nit`

```suggestion
	 * @param mixed $order The order Id or order object.
```

Aligning is necessary only with lines right below or after, right? In this case, there is a new-line.
Not important though as PHP linter does not complain.

File: includes/class-wc-payments-order-service.php
Comment (shendy-a8c) [inline includes/class-wc-payments-order-service.php]:
```suggestion
	 * @throws Order_Not_Found_Exception
     * @throws WCPay\Core\Exceptions\Server\Request\Extend_Request_Exception
	 * @throws WCPay\Core\Exceptions\Server\Request\Immutable_Parameter_Exception
	 * @throws WCPay\Core\Exceptions\Server\Request\Invalid_Request_Parameter_Exception
```

I see the above exceptions might be thrown by `$request->send()`.

File: includes/admin/class-wc-payments-admin.php
Comment (shendy-a8c) [inline includes/admin/class-wc-payments-admin.php]: An exception might be thrown here. When it happens, order edit page will show a fatal error as the thrown exception is not caught.

IMO, the dispute notice is not too crucial to the point where when we fail to fetch the dispute data, the order edit page needs to be inaccessible.

What do you think of wrapping this call with a try-catch and log the error so it's not unnoticed?

File: includes/admin/class-wc-payments-admin.php
Comment (shendy-a8c) [inline includes/admin/class-wc-payments-admin.php]: This comment looks misplaced as it mentions about evidence details but the check for that is above it.

File: client/order/index.js
Comment (shendy-a8c) [inline client/order/index.js]: Should this be inclusive?

```suggestion
		if ( 7 >= countdownDays ) {
```

File: client/order/index.js
Comment (shendy-a8c) [inline client/order/index.js]: Should this be `%s`? Or, change the translated strings to use `%d`.

```suggestion
				// Translators: %s is the number of days left to respond to the dispute.
```

File: client/order/index.js
Comment (shendy-a8c) [inline client/order/index.js]:
```suggestion
		if ( 3 >= countdownDays ) {
```

File: client/order/index.js
Comment (shendy-a8c) [inline client/order/index.js]:
```suggestion
		if ( 1 >= countdownDays ) {
```

File: client/order/index.js
Comment (shendy-a8c) [inline client/order/index.js]: Might not be necessary because `urgency` must have been set to `error` in the previous if clause but OK to be explicit here.

**Review by shendy-a8c — CHANGES_REQUESTED** [review]
Looks good and tested well.

I left some minor comments that some might require some changes. The [one about catching an exception](https://github.com/Automattic/woocommerce-payments/pull/6373#discussion_r1259465602) is the most important IMO.

ps:
- When I click the CTA button, I notice the URL changes.

https://github.com/Automattic/woocommerce-payments/assets/73803630/3e0c76e6-c52a-4769-9de2-18039ea1062a

This is because `add_query_arg()` (used in `compose_dispute_url()`) does not URL-encode the parameter values. I'm a bit surprised by it. It's not a problem because the browser will correct it.

- Also, one improvement might be to fetch the dispute data from the front-end, making the order edit page loads a bit faster without the request to fetch the dispute data. After all, the rendering happens fully in the front-end. This will require a lot of change though, especially if there is no end-point yet to fetch dispute from an order, so I'm not suggesting to pursue it.

**Comment by shendy-a8c** [comment]
The label `pr: needs author reply` does not take account the possibility that a PR might be handed over to someone else =)

File: client/order/index.js
Comment (Jinksi) [reply client/order/index.js]: This is using `>` since the `countdownDays` will equal 6 if within 7 days.

E.g. 7 days is 168 hours. 167 hours (within 7 days) = 6.95 days = `Math.floor(6.95) = 6 countdownDays`.

This can be tested by setting `disputeData.dueBy = moment().add( 167, 'hours' ).unix();`

File: client/order/index.js
Comment (Jinksi) [reply client/order/index.js]: This is using `>` since the `countdownDays` will equal 3 if within 3 days / 72 hours.

E.g. 3 days is 72 hours. 71 hours (within 3 days) = 2.95 days = `Math.floor(2.95) = 2 countdownDays`.

This can be tested by setting `disputeData.dueBy = moment().add( 71, 'hours' ).unix();`

File: client/order/index.js
Comment (Jinksi) [reply client/order/index.js]: This is using `>` since we only want this to resolve `true` if within 24h.
24+ hours would result in `countdownDays = 1` after `Math.floor` which we want to resolve as `false` (not within 24h).

E.g. 23 hours = 0.95 days = `Math.floor(0.95) = 0 countdownDays`.

This can be tested by setting `disputeData.dueBy = moment().add( 23, 'hours' ).unix();`

File: client/order/index.js
Comment (Jinksi) [reply client/order/index.js]: Yeh I decided to leave this to show the reader that this is explicitly defined for both 72h and 24h cases.

File: includes/admin/class-wc-payments-admin.php
Comment (Jinksi) [inline includes/admin/class-wc-payments-admin.php:696]: This shouldn't risk a fatal error, since it is just a wrapped call to `order->get_meta()`.

File: includes/admin/class-wc-payments-admin.php
Comment (Jinksi) [reply includes/admin/class-wc-payments-admin.php]: This should be resolved in aab6e0c since this method has been removed and this logic is now performed on the frontend.

File: includes/admin/class-wc-payments-admin.php
Comment (Jinksi) [reply includes/admin/class-wc-payments-admin.php]: This should be resolved in aab6e0c since this method has been removed and this logic is now performed on the frontend.

**Comment by Jinksi** [comment]
@shendy-a8c Thanks for the review! I've addressed your comments – in particular, I've moved the dispute-fetching logic from PHP to JS, to improve the initial order page load speed in aab6e0c. 

I agree, I don't think it makes sense to reduce the performance for EVERY order page load when a dispute on an order is a relatively rare case and the render must wait for JS to load.

I was able to use the `useCharge` hook to fetch the same data, simple calling `get_charge_id_for_order()` in PHP which is a much quicker order meta lookup.

File: includes/class-wc-payments-order-service.php
Comment (Jinksi) [inline includes/class-wc-payments-order-service.php]: This can be changed back to `private` since it is no longer being used in this PR.
```suggestion
	private function compose_dispute_url( $dispute_id ) {
```

File: client/order/index.js
Comment (Jinksi) [inline client/order/index.js:129]: The logic below is almost identical to before the changes in aab6e0c where I've moved this to a separate `<DisputeNotice />` component to improve readability.

The data from `useCharge` is almost identical to what was previously provided by PHP in `disputeData`, so not many changes were required.

**Review by shendy-a8c — APPROVED** [review]
Looks good! Everything still functions as before. Very nice that you implemented the fetching charge and dispute data asynchronously from the front-end and in a quick turn-around too.

**Pull request merged**
